client/sql/dataframe.go

```diff
@@ -52,11 +66,22 @@ type RangePartitionColumn struct {
 
 // dataFrameImpl is an implementation of DataFrame interface.
 type dataFrameImpl struct {
-	sparkSession *sparkSessionImpl
+	sparkSession sparkExecutor
```

Comment:
shouldn't the context be a similar member here? I don't have enough Go skills to reason about this, but I'm wondering how common this is to pass the context through every public method.

Reply:
If you want a peek down the rabbit hole wrt context: https://go.dev/blog/context. It is explicit (and therefore somewhat verbose) but the pattern enables correctness in all the corner cases that always end up causing problems when using ThreadLocal in JVM languages.

It is a standard construct of Golang and this change basically brings us in line with best/common practice

Reply:
I don't really have a strong opinion; I'm just looking at this from an API-style perspective. 

> At Google, we require that Go programmers pass a Context parameter as the first argument to every function on the call path between incoming and outgoing requests. This allows Go code developed by many different teams to interoperate well. It provides simple control over timeouts and cancellation and ensures that critical values like security credentials transit Go programs properly.

Now, my question is if you have Python code like this:

```python
df = spark.table("mytable")
df = df.filter(col("id") > 10)
df.collect()
```

When would you actually have to pass the context. In abbreviated non-go syntax:

```golang
ctx := context.Background()
df_tab := spark.table("mytable")
df_filt := df.filter(col("id")).gt(10))
res, err := df_filt.collect(ctx)
```

This means the context is explicitly passed for all actions on the DF, but not for local operations. But now, what if `filter` needs to retrieve the schema first. Then it becomes

```golang
ctx := context.Background()
df_tab := spark.table("mytable")
df_filt := df.filter(ctx, col("id")).gt(10))
res, err := df_filt.collect(ctx)
```

But isn't this then very verbose? The DF can conceptually not live without the context either. The cancellation and processing properties are still tied to the DF itself.

I'm not experienced enough to have a good understanding where the public usage of context begins and ends.

Reply:
Yes, that is correct. You will pass an extra parameter and in a sense that is verbose. In practice, this is what every Go programmer out there will expect. The strength of Go is not in its lack of verbosity, it is in how it makes all flows, including contexts and error flows explicit, forcing engineers to think about correctness of all paths. That is why Go is particularly well-suited for programs where a high degree of correctness is needed for the runtime behavior, most prominently in infrastructure and platform code. That is also why it doesn't have unchecked exceptions (or even a try-catch construct for checked ones) but forces you to handle all errors in the ordinary flow, every time.

I have only added context params where they are needed because of outgoing calls, minimizing the number of params that need to be passed. You would, however, also use a context in local computation if you wanted a timeout, e.g., a timeout for how long a filter operation can take.